In [1]:
# Analysis hacker_news data

In [1]:
from csv import reader
import datetime as dt

In [2]:
with open("hacker_news.csv",'r') as fh:
    hn=list(reader(fh))

In [3]:
hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

In [5]:
header=hn[0]
print(header)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [8]:
## remove header
hn=hn[1:]

## Find required data/entries
Since we're only concerned with post titles beginning with Ask HN or Show HN, we'll create new lists of lists containing just the data for those titles.

In [4]:
ask_posts=[]
show_posts=[]
other_posts=[]

In [5]:
for row in hn:
    title=row[1].lower()
    
    if title.startswith("ask"):
        ask_posts.append(row)
    elif title.startswith("show"):
        show_posts.append(row)
    else:
        other_posts.append(row)

In [6]:
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1756
1164
17181


In [7]:
## Find total comments of two type of posts


In [8]:
total_ask_comments=0
for row in ask_posts:
    n_comments=row[4]
    total_ask_comments+=int(n_comments)

In [9]:
avg_ask_comments=total_ask_comments/len(ask_posts)
print(avg_ask_comments)

14.203302961275627


In [10]:
total_show_comments=0
for row in show_posts:
    n_comments=row[4]
    total_show_comments+=int(n_comments)

In [11]:
avg_show_comments=total_show_comments/len(show_posts)
print(avg_show_comments)

10.323024054982818


## 
you should've determined that, on average, ask posts receive more comments than show posts.

Next, we'll determine if ask posts created at a certain time are more likely to attract comments. 

In [12]:
result_list=[]

In [13]:
for row in ask_posts:
    result_list.append([row[6],int(row[4])])

In [26]:
counts_by_hour={}
comments_by_hour={}

In [27]:
time_format="%m/%d/%Y %H:%M"
for row in result_list:
    time=row[0]
    time_obj=dt.datetime.strptime(time,time_format)
    hour=time_obj.hour
    if hour in counts_by_hour:
        counts_by_hour[hour]+=1
        comments_by_hour[hour]+=row[1]
    else:
        counts_by_hour[hour]=1
        comments_by_hour[hour]=row[1]
    

In [28]:
avg_by_hour=[]
for hour in comments_by_hour:
    avg_by_hour.append([hour,comments_by_hour[hour]/counts_by_hour[hour]])
    

In [29]:
avg_by_hour

[[9, 5.5777777777777775],
 [13, 14.581395348837209],
 [10, 13.233333333333333],
 [14, 12.909090909090908],
 [16, 17.880733944954127],
 [23, 10.185714285714285],
 [12, 9.41095890410959],
 [17, 11.46],
 [15, 38.5948275862069],
 [21, 16.009174311926607],
 [20, 21.525],
 [2, 23.810344827586206],
 [18, 13.20183486238532],
 [3, 8.345454545454546],
 [5, 10.08695652173913],
 [19, 11.5625],
 [1, 11.383333333333333],
 [22, 6.746478873239437],
 [8, 10.25],
 [4, 7.0625],
 [0, 8.127272727272727],
 [6, 9.022727272727273],
 [7, 7.852941176470588],
 [11, 11.051724137931034]]

## Sorting and printing list of lists

In [30]:
swap_avg_by_hour=[]

In [31]:
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])
print(swap_avg_by_hour)

[[5.5777777777777775, 9], [14.581395348837209, 13], [13.233333333333333, 10], [12.909090909090908, 14], [17.880733944954127, 16], [10.185714285714285, 23], [9.41095890410959, 12], [11.46, 17], [38.5948275862069, 15], [16.009174311926607, 21], [21.525, 20], [23.810344827586206, 2], [13.20183486238532, 18], [8.345454545454546, 3], [10.08695652173913, 5], [11.5625, 19], [11.383333333333333, 1], [6.746478873239437, 22], [10.25, 8], [7.0625, 4], [8.127272727272727, 0], [9.022727272727273, 6], [7.852941176470588, 7], [11.051724137931034, 11]]


Use the sorted() function to sort swap_avg_by_hour in descending order. Since the first column of this list is the average number of comments, sorting the list will sort by the average number of comments.

In [32]:
sorted_swap=sorted(swap_avg_by_hour,reverse=True)

In [34]:
print("Top 5 Hours for Ask Posts Comments")

Top 5 Hours for Ask Posts Comments


Loop through each average and each hour (in this order) in the first five lists of sorted_swap.

Use the str.format() method to print the hour and average in the following format: 15:00: 38.59 average comments per post.
- To format the hours, use the datetime.strptime() constructor to return a datetime object and then use the strftime() method to specify the format of the time.
- To format the average, you can use {:.2f} to indicate that just two decimal places should be used.

In [35]:
sorted_swap[1]

[23.810344827586206, 2]

In [39]:
for row in sorted_swap:
    time=dt.datetime.strptime(str(row[1]),"%h")
    formated_time=time.strftime("%H:%M")
    print("{time}: {num} average comments per post".format(time=formated_time,num=row[0]))

ValueError: 'h' is a bad directive in format '%h'